Panic: close of closed channel

Got into this issue today but it's not easy to reproduce. The current code should handle channel close more carefully.

```
Enter file path to send: 2019/08/17 07:09:23 WRONG NODE TO CONNECT
2019/08/17 07:09:24 write tcp 157.230.136.253:45156->221.235.118.49:30002: use of closed network connection
2019/08/17 07:09:25 write tcp 157.230.136.253:45158->221.235.118.49:30002: use of closed network connection
2019/08/17 07:09:28 write tcp 157.230.136.253:45160->221.235.118.49:30002: use of closed network connection

...

2019/08/17 07:11:26 write tcp 157.230.136.253:45320->221.235.118.49:30002: use of closed network connection
2019/08/17 07:11:29 write tcp 157.230.136.253:45322->221.235.118.49:30002: use of closed network connection
2019/08/17 07:11:31 write tcp 157.230.136.253:45324->221.235.118.49:30002: use of closed network connection
2019/08/17 07:11:36.439565 [ERROR] GID 6701, POST request: Post http://mainnet-seed-0015.nkn.org:30003: net/http: request canceled (Client.Timeout exceeded while awaiting headers)

2019/08/17 07:11:36 unexpected end of JSON input
panic: close of closed channel

goroutine 6701 [running]:
github.com/nknorg/nkn-sdk-go.(*Client).connect.func2.1(0xc0000dd450)
	/root/go/src/github.com/nknorg/nkn-sdk-go/client.go:143 +0x32
github.com/nknorg/nkn-sdk-go.(*Client).connect.func2(0xc0000dd450, 0xc00042f080)
	/root/go/src/github.com/nknorg/nkn-sdk-go/client.go:250 +0xb9
created by github.com/nknorg/nkn-sdk-go.(*Client).connect
	/root/go/src/github.com/nknorg/nkn-sdk-go/client.go:141 +0x1d0
exit status 2
```

--- a/client.go
+++ b/client.go
@@ -44,7 +44,6 @@ type Client struct {
 	Address   string
 	addressId []byte
 	urlString string
-	closed    bool
 	OnConnect chan struct{}
 	OnMessage chan *pb.InboundMessage
 	OnBlock   chan *BlockInfo
@@ -52,8 +51,9 @@ type Client struct {
 	nodeInfo          *NodeInfo
 	sigChainBlockHash string
 
-	sync.Mutex
-	conn *websocket.Conn
+	sync.RWMutex
+	closed bool
+	conn   *websocket.Conn
 }
 
 type NodeInfo struct {
@@ -125,7 +125,7 @@ func (c *Client) connect(retry uint32) error {
 		return conn, err
 	}()
 
-	if err != nil && !c.closed {
+	if err != nil && !c.IsClosed() {
 		log.Println(err)
 
 		time.Sleep(c.config.ReconnectInterval * time.Second)
@@ -139,13 +139,7 @@ func (c *Client) connect(retry uint32) error {
 	c.OnBlock = make(chan *BlockInfo, c.config.BlockChanLen)
 
 	go func() {
-		defer func() {
-			close(c.OnConnect)
-			close(c.OnMessage)
-			close(c.OnBlock)
-
-			_ = c.conn.Close()
-		}()
+		defer c.Close()
 
 		err := func() error {
 			req := make(map[string]interface{})
@@ -244,7 +238,7 @@ func (c *Client) connect(retry uint32) error {
 			log.Println(err)
 		}
 
-		if !c.closed {
+		if !c.IsClosed() {
 			defer c.connect(0)
 		}
 	}()
@@ -432,9 +426,20 @@ func (c *Client) Publish(topic string, bucket uint32, payload []byte, MaxHolding
 	return c.Send(dests, payload, MaxHoldingSeconds...)
 }
 
+func (c *Client) IsClosed() bool {
+	c.RLock()
+	defer c.RUnlock()
+	return c.closed
+}
+
 func (c *Client) Close() {
+	c.Lock()
+	defer c.Unlock()
 	if !c.closed {
 		c.closed = true
-		_ = c.conn.Close()
+		close(c.OnConnect)
+		close(c.OnMessage)
+		close(c.OnBlock)
+		c.conn.Close()
 	}
 }

